### PR TITLE
[AutoDiff] delete SILDifferentiableAttribute

### PR DESCRIPTION
**Pull request by marcrasi**
Deletes SILDifferentiableAttribute (TF-898), and makes differentiation directly canonicalize SILDifferentiabilityWitnesses (TF-1024).

Since this is a pretty "dangerous" change, I'll run swift-apis and swift-models tests on it.

### CODE REVIEW

**Review by rxwei — APPROVED** [review]
How can I not love this?  Thank you for deleting this!

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Review by dan-zheng — APPROVED** [review]
👍👍

**Comment by marcrasi** [comment]
The tests failed under optimized mode because this change made all the derivative functions hidden, and the devirtualization pass tries to function_ref. So I reverted the part that makes all the derivative functions hidden, which should fix that.

Also swift-apis failed. I'll rerun it with my new commit, but I doubt that'll fix it. I'll investigate this failure more tomorrow.
```
Step 17/34 : RUN /swift-tensorflow-toolchain/usr/bin/swift test
 ---> Running in 61d938d88942
[1/13] Compiling Tensor Random.swift
[3/14] Merging module Experimental
[4/16] Wrapping AST for Experimental for debugging
[5/16] Compiling ExperimentalTests XCTestManifests.swift
[6/17] Merging module Tensor
[7/21] Wrapping AST for Tensor for debugging
[8/33] Linking libTensor.so
[9/71] Compiling TensorFlow Exports.swift
[10/71] Compiling TensorFlow Freezable.swift
[11/71] Compiling TensorFlow Optimizer.swift
[12/71] Compiling TensorFlow Execution.swift
[13/71] Compiling TensorFlow LazyTensorContext.swift
[14/71] Compiling TensorFlow Threading.swift
[15/71] Compiling TensorFlow Utilities.swift
[16/71] Compiling TensorFlow Context.swift
[17/71] Compiling TensorFlow ArrayOps.swift
[18/71] Compiling TensorFlow StdlibExtensions.swift
[19/71] Compiling TensorFlow SGD.swift
[20/71] Compiling TensorFlow Runtime.swift
[21/71] Compiling TensorFlow Serialization.swift
[22/71] Compiling TensorFlow LazyTensorTraceCache.swift
[23/71] Compiling TensorFlow PythonConversion.swift
[24/71] Compiling TensorFlow TensorHandle.swift
[25/71] Compiling TensorFlow TensorShape.swift
[26/71] Compiling TensorFlow LazyTensorTFFunctionBuilder.swift
[27/71] Compiling TensorFlow LazyTensorTrace.swift
[28/71] Compiling TensorFlow Comparison.swift
[29/71] Compiling TensorFlow RawOpsGenerated.swift
[30/71] Compiling TensorFlow EagerExecution.swift
[31/71] Compiling TensorFlow Dataset.swift
[33/71] Compiling TensorFlow DataTypes.swift
[34/71] Compiling TensorFlow DifferentialOperators.swift
[35/71] Compiling TensorFlow BackwardsCompatibility.swift
[36/72] Merging module TensorTests
[37/72] Wrapping AST for TensorTests for debugging
[39/73] Merging module ExperimentalTests
[40/73] Wrapping AST for ExperimentalTests for debugging
[41/73] Compiling TensorFlow LazyTensorOperation.swift
[42/73] Compiling TensorFlow LazyTensorShapeInference.swift
[43/73] Compiling TensorFlow ShapedArray.swift
[44/73] Compiling TensorFlow StringTensor.swift
[45/73] Compiling TensorFlow Sequential.swift
[46/73] Compiling TensorFlow Embedding.swift
[47/73] Compiling TensorFlow Initialization.swift
[48/73] Compiling TensorFlow Tensor.swift
[49/73] Compiling TensorFlow TensorGroup.swift
[52/73] Compiling TensorFlow Initializers.swift
[53/73] Compiling TensorFlow Layer.swift
[54/73] Compiling TensorFlow MomentumBased.swift
[55/73] Compiling TensorFlow NN.swift
[56/73] Compiling TensorFlow Dense.swift
[57/73] Compiling TensorFlow Dropout.swift
[58/73] Compiling TensorFlow Upsampling.swift
[60/73] Compiling TensorFlow TFTensorOperation.swift
[61/73] Compiling TensorFlow Recurrent.swift
[62/73] Compiling TensorFlow Basic.swift
[63/73] Compiling TensorFlow Loss.swift
[64/73] Compiling TensorFlow Normalization.swift
[65/73] Compiling TensorFlow Pooling.swift
[66/73] Compiling TensorFlow Convolutional.swift
[67/73] Compiling TensorFlow Core.swift
[68/73] Compiling TensorFlow Math.swift
[69/74] Merging module TensorFlow
[70/81] Wrapping AST for TensorFlow for debugging
[71/100] Compiling TensorFlowTests LazyTensorTestHelper.swift
[72/100] Linking libTensorFlow.so
[73/100] Compiling TensorFlowTests ContextTests.swift
[74/100] Compiling TensorFlowTests RuntimeTests.swift
[75/100] Compiling TensorFlowTests ComparisonTests.swift
[76/100] Compiling TensorFlowTests LazyTensorTFFunctionBuilderTests.swift
[77/100] Compiling TensorFlowTests LazyTensorTraceCacheTests.swift
[78/100] Compiling TensorFlowTests Helpers.swift
[79/100] Compiling TensorFlowTests LazyTensorShapeInferenceTests.swift
[80/100] Compiling TensorFlowTests TensorTests.swift
[81/100] Compiling TensorFlowTests LazyTensorHandleTests.swift
[82/100] Compiling TensorFlowTests LazyTensorTraceTests.swift
[83/100] Compiling TensorFlowTests LazyTensorOperationTests.swift
[84/100] Compiling TensorFlowTests LazyTensorExplicitTraceTests.swift
[85/100] Compiling TensorFlowTests LazyTensorEvaluationTests.swift
[86/100] Compiling TensorFlowTests TensorGroupTests.swift
[87/100] Compiling TensorFlowTests InitializerTests.swift
[88/100] Compiling TensorFlowTests FreezableTests.swift
[89/100] Compiling TensorFlowTests DatasetTests.swift
[90/100] Compiling TensorFlowTests TrivialModelTests.swift
/swift-apis/Tests/TensorFlowTests/TrivialModelTests.swift:48:44: warning: 'gradient(in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level
function 'TensorFlow.gradient(at:in:)' instead
                let 𝛁model = classifier.gradient { classifier -> Tensor<Float> in
[91/100] Compiling TensorFlowTests XCTestManifests.swift
[92/100] Compiling TensorFlowTests SequentialTests.swift
/swift-apis/Tests/TensorFlowTests/SequentialTests.swift:50:39: warning: 'gradient(in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level
function 'TensorFlow.gradient(at:in:)' instead
                let 𝛁model = model.gradient { model -> Tensor<Float> in
[93/100] Compiling TensorFlowTests MathTests.swift
[94/100] Compiling TensorFlowTests LossTests.swift
[95/100] Compiling TensorFlowTests BasicTests.swift
/swift-apis/Tests/TensorFlowTests/OperatorTests/BasicTests.swift:79:23: warning: 'gradient(in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level
function 'TensorFlow.gradient(at:in:)' instead
        let grads = x.gradient { a -> Tensor<Float> in
                      ^
/swift-apis/Tests/TensorFlowTests/OperatorTests/BasicTests.swift:89:23: warning: 'gradient(in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level
function 'TensorFlow.gradient(at:in:)' instead
        let grads = x.gradient { a -> Tensor<Float> in
                      ^
/swift-apis/Tests/TensorFlowTests/OperatorTests/BasicTests.swift:99:23: warning: 'gradient(in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level
function 'TensorFlow.gradient(at:in:)' instead
        let grads = x.gradient { a -> Tensor<Float> in
                      ^
/swift-apis/Tests/TensorFlowTests/OperatorTests/BasicTests.swift:109:23: warning: 'gradient(in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level
function 'TensorFlow.gradient(at:in:)' instead
        let grads = x.gradient { a -> Tensor<Float> in
                      ^
[96/100] Compiling TensorFlowTests LayerTests.swift
/swift-apis/Tests/TensorFlowTests/LayerTests.swift:51:43: warning: 'gradient(in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level
function 'TensorFlow.gradient(at:in:)' instead
                    let 𝛁model = model.gradient { model -> Tensor<Float> in
/swift-apis/Tests/TensorFlowTests/LayerTests.swift:1145:36: warning: 'valueWithPullback(at:in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level function 'Swift.valueWithPullback(at:_:in:)' instead
            let (outputs, _) = rnn.valueWithPullback(at: inputs) { rnn, inputs in
                                   ^
/swift-apis/Tests/TensorFlowTests/LayerTests.swift:1176:40: warning: 'valueWithPullback(at:in:)' is deprecated: Method-style differential operators are deprecated and will be removed; use top-level function 'Swift.valueWithPullback(at:_:in:)' instead
                let (outputs, _) = rnn.valueWithPullback(at: inputs) { rnn, inputs in
                                       ^
[97/100] Compiling TensorFlowTests TensorAutoDiffTests.swift
[98/101] Merging module TensorFlowTests
[99/102] Wrapping AST for TensorFlowTests for debugging
[100/102] Compiling TensorFlowPackageTests LinuxMain.swift
[101/103] Merging module TensorFlowPackageTests
[102/103] Wrapping AST for TensorFlowPackageTests for debugging
/usr/bin/ld.gold: error: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlowTests.build/LossTests.swift.o: multiple definition of 'AD__$s10TensorFlow0A0V8reshaped2toACyxGAA0A5ShapeV_tF_PUSRS10TensorFlow0aB13FloatingPointRzl'
/usr/bin/ld.gold: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlow.build/Operators/Basic.swift.o: previous definition here
/usr/bin/ld.gold: error: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlowTests.build/LossTests.swift.o: multiple definition of 'AD__$s10TensorFlow0A0V8reshaped7toShapeACyxGACys5Int32VG_tF_PUSRS10TensorFlow0aB13FloatingPointRzl'
/usr/bin/ld.gold: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlow.build/Operators/Basic.swift.o: previous definition here
/usr/bin/ld.gold: error: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlowTests.build/LossTests.swift.o: multiple definition of 'AD__$s10TensorFlow0A0V9flattenedACyxGyF_PSRS10TensorFlow0aB13FloatingPointRzl'
/usr/bin/ld.gold: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlow.build/Operators/Basic.swift.o: previous definition here
/usr/bin/ld.gold: error: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlowTests.build/LossTests.swift.o: multiple definition of 'AD__$s10TensorFlow0A0VAASjRzrlE3sum13squeezingAxesACyxGACys5Int32VG_tF_PUSRS10TensorFlow0aB13FloatingPointRzl'
/usr/bin/ld.gold: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlow.build/Operators/Math.swift.o: previous definition here
/usr/bin/ld.gold: error: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlowTests.build/LossTests.swift.o: multiple definition of 'AD__$s10TensorFlow0A0VAASjRzrlE3sum13squeezingAxesACyxGSaySiG_tF_PUSRS10TensorFlow0aB13FloatingPointRzl'
/usr/bin/ld.gold: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlow.build/Operators/Math.swift.o: previous definition here
/usr/bin/ld.gold: error: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlowTests.build/LossTests.swift.o: multiple definition of 'AD__$s10TensorFlow0A0VAASjRzrlE3sum13squeezingAxesACyxGSid_tF_PUSRS10TensorFlow0aB13FloatingPointRzl'
/usr/bin/ld.gold: /swift-apis/.build/x86_64-unknown-linux-gnu/debug/TensorFlow.build/Operators/Math.swift.o: previous definition here
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: link command failed with exit code 1 (use -v to see invocation)
```

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: lib/SILGen/SILGenPoly.cpp [hunk 1]
```diff
@@ -3714,7 +3714,7 @@ SILGenModule::getOrCreateAutoDiffDerivativeReabstractionThunk(
   // This thunk is publicly exposed and cannot be transparent.
   // Instead, mark it as "always inline" for optimization.
   auto *thunk = fb.getOrCreateFunction(
-      loc, name, SILLinkage::Hidden, origDerivativeFnType, IsBare,
+      loc, name, original->getLinkage(), origDerivativeFnType, IsBare,
```
Comment (rxwei) [inline lib/SILGen/SILGenPoly.cpp]: The linkage should be the same as the differentiability witness, not the original function.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: lib/SILGen/SILGenPoly.cpp (hunk 1 above)
Comment (marcrasi) [reply lib/SILGen/SILGenPoly.cpp]: It was already using the original function linkage before this PR, so I will leave it as it is for now.

There are a big number of problems caused by incorrect linkage piling up, so I do want to carefully think through the correct linkage for everything and create a PR that fixes all the linkage soon. I have filed a ticket for that: https://bugs.swift.org/browse/TF-1027

**Comment by marcrasi** [comment]
Wow, my last commit somehow fixed the swift-apis test issue!? That's weird. I'll investigate the reason as part of my work on https://bugs.swift.org/browse/TF-1027.

**Comment by rxwei** [comment]
I suspect it's because the original function had `Shared` linkage (`internal` access), which allowed duplicate symbols between your test modules.

**Pull request merged**